StreamChatClient_v3/Controllers/Controller.swift

```diff
@@ -6,17 +6,31 @@ import Foundation
 
 /// The base class for all controllers. Not meant to be used directly.
 public class Controller {
+
     /// Describes the possible states of Controller
     public enum State: Equatable {
-        /// The controller is idle. Call `startUpdating` to start listening for changes.
-        case idle
-        
-        /// The controller is active and is listening to changes.
-        case active
+        /// The controller is inactive. Call `startUpdating` to start listening for changes.
+        case inactive
+        /// The controller already fetched local data if any anf remote data fetching is in progress.
+        case localDataFetched
+        /// The controller fetched remote data.
+        case remoteDataFetched
+        /// The controller failed to fetch remote data.
+        case remoteDataFetchFailed(ClientError)
     }
     
     /// The current state of the Controller.
-    public internal(set) var state: State = .idle
+    public internal(set) var state: State = .inactive {
+        didSet {
+            callback { [weak self] in
+                guard let self = self else { return }
+                self.stateDelegate?.controller(self, didChangeState: self.state)
+            }
```

Comment:
Hmm, I think `self` should be referenced strongly here. We want to be sure the delegate gets called for possible user-code cleanups.